quicklogic/qlf_k4n8/utils/create_lib.py

```diff
@@ -0,0 +1,416 @@
+"""
+import argparse
+import sys
+import csv
+import os
+from collections import namedtuple
+from datetime import date
+import lxml.etree as ET
+
+# =============================================================================
+
+# =============================================================================
+"""
+Pin properties
+name - pin_name
+dir - direction (input/output)
+used - specify 'yes' if user is using the pin else specify 'no'
+clk - specify associate clock
+"""
+PinData = namedtuple("PinData", "name dir used clk")
+
+# =============================================================================
+
+
+def main():
+    """
+    Creates a device library file by getting data from given csv and xml file
+    """
+    parser = argparse.ArgumentParser(
+        description='Creates a device library file.'
+    )
+    parser.add_argument(
+        "--lib",
+        "-l",
+        "-L",
+        type=str,
+        default="qlf_k4n8.lib",
+        help='The output device lib file'
+    )
+    parser.add_argument(
+        "--lib_name",
+        "-n",
+        "-N",
+        type=str,
+        required=True,
+        help='Specify library name'
+    )
+    parser.add_argument(
+        "--cell_name",
+        "-m",
+        "-M",
+        type=str,
+        required=True,
+        help='Specify cell name'
+    )
+    parser.add_argument(
+        "--csv",
+        "-c",
+        "-C",
+        type=str,
+        required=True,
+        help='Input pin-map csv file'
+    )
+    parser.add_argument(
+        "--xml",
+        "-x",
+        "-X",
+        type=str,
+        required=True,
+        help='Input interface-mapping xml file'
+    )
+
+    args = parser.parse_args()
+
+    csv_pin_data = set()
+    assoc_clk = dict()
+    with open(args.csv, newline='') as csvfile:
+        reader = csv.DictReader(csvfile)
+        for row in reader:
+            csv_pin_data.add(row['port_name'])
+            if row['Associated Clock'] is not None:
+                assoc_clk[row['port_name']] = row['Associated Clock'].strip()
+
+    port_names = parse_xml(args.xml)
+    create_lib(
+        port_names, csv_pin_data, args.lib_name, args.lib, args.cell_name,
+        assoc_clk
+    )
+
+
+# =============================================================================
+def create_lib(
+        port_names, csv_pin_data, lib_name, lib_file_name, cell_name, assoc_clk
+):
+    """
+    Create lib file
+    """
+    # Read header template file and populate lib file with this data
+    curr_dir = os.path.dirname(os.path.abspath(__file__))
+
+    header_templ_file = os.path.join(
+        curr_dir, '../devices/umc22/lib_header_template.txt'
+    )
+    in_str = open(header_templ_file, 'r').read()
+    curr_str = in_str.replace("@lib_name@", lib_name)
+    today = date.today()
+    curr_date = today.strftime("%B %d, %Y")
+    str_rep_date = curr_str.replace("@Date@", curr_date)
+    str1 = str_rep_date.replace("@cell_name@", cell_name)
```

Comment:
Instead of doing `str.replace` with custom placeholder strings you may process the whole text with `str.format`. It already has a mechanism for what you want and is easily extendable. This is just a suggestion.

Reply:
I cannot use str.format in this case because curly braces are even used in the text and that conflicts with keywords specification for format.